docs/decisions/0033-rag-in-sk.md

```diff
@@ -0,0 +1,297 @@
+
+```json
+{
+    // Other properties...
+    "status": {
+        "ready": false,
+        "state": "Initializing"
+    }
+}
+```
+
+In this case, it's impossible to insert a record to database immediately, so HTTP polling or similar mechanism should be implemented to cover this scenario.
+
+### MemoryRecord as storage schema
+
+`IMemoryStore` interface uses `MemoryRecord` class as storage schema in vector DB. This means that `MemoryRecord` properties should be aligned to all possible connectors. As soon as developers will use this schema in their databases, any changes to schema may break the application, which is not flexible approach.
+
+`MemoryRecord` contains property `ReadOnlyMemory<float> Embedding` for embeddings and `MemoryRecordMetadata Metadata` for embeddings metadata. `MemoryRecordMetadata` contains properties like:
+
+- `string Id` - unique identifier.
+- `string Text` - data-related text.
+- `string Description` - optional title describing the content.
+- `string AdditionalMetadata` - field for saving custom metadata with a record.
+
+Since `MemoryRecord` and `MemoryRecordMetadata` are not sealed classes, it should be possible to extend them and add more properties as needed. Although, current approach still forces developers to have specific base schema in their vector DBs, which ideally should be avoided. Developers should have the ability to work with any schema of their choice, which will cover their business scenarios (similarly to Code First approach in Entity Framework).
+
+### TextMemoryPlugin
+
+TextMemoryPlugin contains 4 Kernel functions:
+
+- `Retrieve` - returns concrete record from DB by key.
+- `Recall` - performs vector search and returns multiple records based on relevance.
+- `Save` - saves record in vector DB.
+- `Remove` - removes record from vector DB.
+
+All functions can be called directly from prompt. Moreover, as soon as these functions are registered in Kernel and Function Calling is enabled, LLM may decide to call specific function to achieve provided goal.
+
+`Retrieve` and `Recall` functions are useful to provide some context to LLM and ask a question based on data, but functions `Save` and `Remove` perform some manipulations with data in vector DB, which could be unpredicted or sometimes even dangerous (there should be no situations when LLM decides to remove some records, which shouldn't be deleted).
+
+## Decision Drivers
+
+1. All manipulations with data in Semantic Kernel should be safe.
+2. There should be a clear way(s) how to use RAG pattern in Semantic Kernel.
+3. Abstractions should not block developers from using vector DB of their choice with functionality, that cannot be achieved with provided interfaces or data types.
```

Comment:
I think that's true, but it's also important to realize that any abstraction here is going to block *some* subset of vector database functionality; that is, even if we evolve the IMemoryStorage interface to support a more powerful/flexible data scheme, there are still many other aspects beyond that which won't be accessible when using the abstraction. In other words, to get the full power/flexibility/feature-set of your specific chosen vector db solution, you most likely will need to drop the abstraction and use its SDK directly.

Just saying this because it's important to not create hopes for a better abstraction that will fully "solve" this problem. For comparison, relational databases seem generally much more similar to each other than vector databases are (there's a SQL standard etc.), and even there the abstraction (ADO.NET in the .NET case) doesn't actually allow fully working with them without things leaking and certain features being blocked.

Reply:
That's a good example. I agree, as discussed in other comments, it sounds like keeping an abstraction should allow developers to handle basic scenarios quickly, while for full flexibility it will be possible to use vector DB directly (e.g. as plugin).

Reply:
I think at the very least, however, that we should provide an ability within our connectors to provide a custom schema. In customer conversations, I've run into quite a few scenarios where they just want to query an index they've already created previously within Semantic Kernel.

I think this is important because it allows devs coming from other frameworks (e.g., LangChain) to quickly test our "getting started" abstractions without having to rebuild their index.

Reply:
Sure, that may indeed be a thing that's too limiting/restrictive with the current IMemoryStore design - I don't necessarily disagree.